core/chains/evm/abi/selector_parser.go

```diff
@@ -0,0 +1,249 @@
+		return nil, fmt.Errorf("failed to parse selector '%s': unexpected string '%s'", unescapedSelector, rest)
+	}
+	return result, nil
+}
+
+func assembleArg(name string, arg any) (abi.ArgumentMarshaling, error) {
+	if s, ok := arg.(string); ok {
+		return abi.ArgumentMarshaling{Name: name, Type: s, InternalType: s, Components: nil, Indexed: false}, nil
+	} else if components, ok := arg.([]interface{}); ok {
+		subArgs, err := assembleArgs(components)
+		if err != nil {
+			return abi.ArgumentMarshaling{}, fmt.Errorf("failed to assemble components: %v", err)
+		}
+		tupleType := "tuple"
+		if len(subArgs) != 0 && subArgs[len(subArgs)-1].Type == "[]" {
+			subArgs = subArgs[:len(subArgs)-1]
+			tupleType = "tuple[]"
+		}
+		return abi.ArgumentMarshaling{Name: name, Type: tupleType, InternalType: tupleType, Components: subArgs, Indexed: false}, nil
+	}
+	return abi.ArgumentMarshaling{}, fmt.Errorf("failed to assemble args: unexpected type %T", arg)
+}
+
+func assembleArgs(args []interface{}) ([]abi.ArgumentMarshaling, error) {
+	arguments := make([]abi.ArgumentMarshaling, 0)
+	for i, arg := range args {
+		// generate dummy name to avoid unmarshal issues
+		name := fmt.Sprintf("name%d", i)
+		arg, err := assembleArg(name, arg)
+		if err != nil {
+			return nil, err
+		}
+		arguments = append(arguments, arg)
+	}
+	return arguments, nil
+}
+
+// ParseSelector converts a method selector into a struct that can be JSON encoded
+// and consumed by other functions in this package.
+// Note, although uppercase letters are not part of the ABI spec, this function
+// still accepts it as the general format is valid.
+func ParseSelector(unescapedSelector string) (abi.SelectorMarshaling, error) {
+	name, rest, err := parseIdentifier(unescapedSelector)
+	if err != nil {
+		return abi.SelectorMarshaling{}, fmt.Errorf("failed to parse selector '%s': %v", unescapedSelector, err)
+	}
+	args := []interface{}{}
+	if len(rest) >= 2 && rest[0] == '(' && rest[1] == ')' {
+		rest = rest[2:]
+	} else {
+		args, rest, err = parseCompositeType(rest)
+		if err != nil {
+			return abi.SelectorMarshaling{}, fmt.Errorf("failed to parse selector '%s': %v", unescapedSelector, err)
+		}
+	}
+	if len(rest) > 0 {
+		return abi.SelectorMarshaling{}, fmt.Errorf("failed to parse selector '%s': unexpected string '%s'", unescapedSelector, rest)
+	}
+
+	// Reassemble the fake ABI and construct the JSON
+	fakeArgs, err := assembleArgs(args)
+	if err != nil {
+		return abi.SelectorMarshaling{}, fmt.Errorf("failed to parse selector: %v", err)
+	}
+
+	return abi.SelectorMarshaling{Name: name, Type: "function", Inputs: fakeArgs}, nil
+}
+
+// ParseSelector converts a method selector into a struct that can be JSON encoded
+// and consumed by other functions in this package.
+// Note, although uppercase letters are not part of the ABI spec, this function
+// still accepts it as the general format is valid.
+func ParseSignature(unescapedSelector string) (abi.SelectorMarshaling, error) {
```

Comment:
Why is this named `unescpaedSelector`? I can't find an "escaped" version. What does it need to be unescaped from?

Reply:
I modified this verbatim from geth (ParseSelector being the original) so I retained the naming

Reply:
it's mostly just because of naming here so the variable names don't conflict:

https://github.com/ethereum/go-ethereum/blob/5d984796afd4aa7c00c6663f4155488a9df73d0e/signer/fourbyte/abi.go#L77-L80